src/Authentication/Common.Authentication/Models/AzureRMProfile.cs

```diff
@@ -0,0 +1,140 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Common.Authentication.Models
+{
+    /// <summary>
+    /// Represents Azure Resource Manager profile structure with default context, environments and token cache.
+    /// </summary>
+    [Serializable]
+    public sealed class AzureRMProfile : IAzureProfile
+    {
+        /// <summary>
+        /// Gets or sets Azure environments.
+        /// </summary>
+        public Dictionary<string, AzureEnvironment> Environments { get; set; }
+
+        /// <summary>
+        /// Gets or sets the token cache contents.
+        /// </summary>
+        public byte[] TokenCache { get; set; }
+
+        /// <summary>
+        /// Gets or sets the default azure context object.
+        /// </summary>
+        public AzureContext DefaultContext { get; set; }
+
+        /// <summary>
+        /// Gets the path of the profile file. 
+        /// </summary>
+        [JsonIgnore]
+        public string ProfilePath { get; private set; }
+
+        private void Load(string path)
+        {
+            this.ProfilePath = path;
+
+            if (!AzureSession.DataStore.DirectoryExists(AzureSession.ProfileDirectory))
+            {
+                AzureSession.DataStore.CreateDirectory(AzureSession.ProfileDirectory);
+            }
+
+            if (AzureSession.DataStore.FileExists(ProfilePath))
+            {
+                string contents = AzureSession.DataStore.ReadFileAsText(ProfilePath);
+                AzureRMProfile profile = JsonConvert.DeserializeObject<AzureRMProfile>(contents);
+                this.DefaultContext = profile.DefaultContext;
+                this.Environments = profile.Environments;
+                this.TokenCache = profile.TokenCache;
+            }
+        }
+
+        /// <summary>
+        /// Creates new instance of AzureRMProfile.
+        /// </summary>
+        public AzureRMProfile()
+        {
+            Environments = new Dictionary<string, AzureEnvironment>(StringComparer.InvariantCultureIgnoreCase);
+
+            // Adding predefined environments
+            foreach (AzureEnvironment env in AzureEnvironment.PublicEnvironments.Values)
+            {
+                Environments[env.Name] = env;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of AzureRMProfile and loads its content from specified path.
+        /// </summary>
+        /// <param name="path">The location of profile file on disk.</param>
+        public AzureRMProfile(string path) : this()
+        {
+            Load(path);
+        }
+
+        /// <summary>
+        /// Writes profile to a specified path.
+        /// </summary>
+        /// <param name="path"></param>
+        public void Save(string path)
```

Comment:
I wonder if we should have a Save overload that takes no parameters, and have this method check the ProfilePath property and throw if it is empty and otherwise simply call Save(this.ProfilePath).  This way callers don't need to know the details of how we are saving the profile path.